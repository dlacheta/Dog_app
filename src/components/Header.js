import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
  <nav className="navbar fixed-top">
    <Link className="navbar-brand" to="/">
      Dog app
    </Link>
    <Link to="/favorites">Favorites</Link>
  </nav>
);

export default Header;
