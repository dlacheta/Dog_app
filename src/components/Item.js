import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toogleFavorites } from '../actions/rootActions';

export const Item = (props) => {
  const { dogData, toogleFavoritesDispatch } = props;
  const startToogleFavorites = (id) => {
    toogleFavoritesDispatch(id);
  };
  return (
    <div className="row">
      {dogData.map(dog => (
        <div key={dog.id} className="img-container d-flex">
          <img
            onClick={() => startToogleFavorites(dog.id)}
            src={dog.src}
            className={`img-responsive rounded ${dog.favorites ? 'favorite' : ''}`}
            alt="dog"
          />
          <i className="material-icons favorites-icon">favorite</i>
        </div>
      ))}
    </div>
  );
};

Item.propTypes = {
  toogleFavoritesDispatch: PropTypes.func.isRequired,
  dogData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      src: PropTypes.string,
      favorites: PropTypes.bool,
    }),
  ),
};

Item.defaultProps = {
  dogData: [],
};

const mapDispatchToProps = dispatch => ({
  toogleFavoritesDispatch: id => dispatch(toogleFavorites(id)),
});

export default connect(
  null,
  mapDispatchToProps,
)(Item);
