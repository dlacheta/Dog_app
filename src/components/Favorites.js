import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Item from './Item';

export const Favorites = (props) => {
  const { dogData } = props;
  return (
    <React.Fragment>
      <div className="alert alert-info">
        Click on the picture to remove dog from Your favorites!
      </div>
      <div className="container-fluid">
        {dogData.length > 0 ? <Item dogData={dogData} /> : "You don't like any dog :("}
      </div>
    </React.Fragment>
  );
};

Favorites.propTypes = {
  dogData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      src: PropTypes.string,
      favorites: PropTypes.bool,
    }),
  ),
};

Favorites.defaultProps = {
  dogData: [],
};

const mapStateToProps = state => ({ dogData: state.dogData.filter(dog => dog.favorites) });
export default connect(mapStateToProps)(Favorites);
