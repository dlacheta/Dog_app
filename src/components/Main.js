import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getDog, toogleLoading, toogleRender } from '../actions/rootActions';
import Item from './Item';

export class Main extends React.Component {
  constructor(props) {
    super(props);
    this.onScroll = this.onScroll.bind(this);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.onScroll);
    const { renderedOnce, getDogDispatch, toogleRenderDispatch } = this.props;
    if (!renderedOnce) {
      getDogDispatch();
      toogleRenderDispatch();
    }
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.onScroll);
  }

  onScroll() {
    const { loadingDog, toogleLoadingDispatch, getDogDispatch } = this.props;
    if (
      window.innerHeight + document.documentElement.scrollTop
        === document.documentElement.offsetHeight
      && !loadingDog
    ) {
      toogleLoadingDispatch();
      getDogDispatch();
    }
  }

  render() {
    const { dogData, loadingDog, error } = this.props;
    return (
      <div>
        <div className="alert alert-info">Click on the picture to add dog to Your favorites!</div>
        <div className="container-fluid">
          {!error ? (
            <Item dogData={dogData} />
          ) : (
            `Unfortunatelly because of "${error}" We cannot provide You cute dogs, please try again later`
          )}
        </div>
        {loadingDog ? (
          <div className="loading-container">
            <div className="loading" />
          </div>
        ) : null}
      </div>
    );
  }
}

Main.propTypes = {
  getDogDispatch: PropTypes.func.isRequired,
  toogleLoadingDispatch: PropTypes.func.isRequired,
  toogleRenderDispatch: PropTypes.func.isRequired,
  loadingDog: PropTypes.bool,
  renderedOnce: PropTypes.bool,
  error: PropTypes.string,
  dogData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      src: PropTypes.string,
      favorites: PropTypes.bool,
    }),
  ),
};

Main.defaultProps = {
  dogData: [],
  loadingDog: true,
  renderedOnce: false,
  error: '',
};

const mapStateToProps = state => ({
  dogData: state.dogData.filter(dog => !dog.favorites),
  loadingDog: state.loadingDog,
  renderedOnce: state.renderedOnce,
  error: state.error,
});
const mapDispachToProps = dispatch => ({
  getDogDispatch: () => dispatch(getDog()),
  toogleLoadingDispatch: () => dispatch(toogleLoading()),
  toogleRenderDispatch: () => dispatch(toogleRender()),
});

export default connect(
  mapStateToProps,
  mapDispachToProps,
)(Main);
