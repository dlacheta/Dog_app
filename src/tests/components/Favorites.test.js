import React from 'react';
import { shallow } from 'enzyme';
import { Favorites } from '../../components/Favorites';

it('Favorites component renders correctly', () => {
  const wrapper = shallow(<Favorites />);
  expect(wrapper).toMatchSnapshot();
});
