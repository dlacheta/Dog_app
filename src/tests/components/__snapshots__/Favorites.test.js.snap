// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    dogData={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="alert alert-info"
        >
          Click on the picture to remove dog from Your favorites!
        </div>,
        <div
          className="container-fluid"
        >
          You don't like any dog :(
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Click on the picture to remove dog from Your favorites!",
          "className": "alert alert-info",
        },
        "ref": null,
        "rendered": "Click on the picture to remove dog from Your favorites!",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "You don't like any dog :(",
          "className": "container-fluid",
        },
        "ref": null,
        "rendered": "You don't like any dog :(",
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="alert alert-info"
          >
            Click on the picture to remove dog from Your favorites!
          </div>,
          <div
            className="container-fluid"
          >
            You don't like any dog :(
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Click on the picture to remove dog from Your favorites!",
            "className": "alert alert-info",
          },
          "ref": null,
          "rendered": "Click on the picture to remove dog from Your favorites!",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "You don't like any dog :(",
            "className": "container-fluid",
          },
          "ref": null,
          "rendered": "You don't like any dog :(",
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
